### PR TITLE
Fix free topology - V5

### PR DESCRIPTION
**Pull request by ranj063**
Only compile tested so far. Please do not merge.

### CODE REVIEW

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -1921,11 +1921,15 @@ static int sof_dai_load(struct snd_soc_component *scomp, int index,
...
-	kfree(spcm);
 
+	/*
+	 * When a new pcm is created, the spcm variable
+	 * that was allocated in sof_dai_load()
+	 * is saved as the private member of pcm_runtime struct.
+	 * So it should be freed during pcm_free() so that it remains
+	 * accessible via the pcm_runtime's private member
+	 * until the pcm is freed.
+	 */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: Aren't we working around a call sequence in the ASoC core that maybe should be changed instead?

File: sound/soc/sof/topology.c [hunk 2]
```diff
@@ -1833,6 +1847,36 @@ static int sof_widget_unload(struct snd_soc_component *scomp,
...
+
+			/* free sroute and its private data */
+			kfree(sroute->private);
+			list_del(&sroute->list);
+			kfree(sroute);
+		}
+
+		/* check sink widget */
+		rwidget = snd_sof_find_swidget(sdev,
+					       (char *)sroute->route.sink);
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: You have a NULL pointer dereference right there in case sroute is freed on line 1863. You can only free the route when both sink and source have been dealt with.

This code is really hard to follow since it's not the same pattern as for loads. Maybe we can rewrite the load part is more contained ways and follow the same pattern for unload.

Comment (ranj063) [reply sound/soc/sof/topology.c]: @plbossart Ahh yes, I should return if the previous if is true. 
Let me look into re-writing the widgets load part but the problem is widget load doesnt quite address routes. Topology parsing happens in phases and the routes come after the widgets are loaded.

**Comment by bardliao** [comment]
@ranj063 I have a suggestion. We can remove ```sof_pcm_free()``` and do all the allocate/free on ```sof_dai_load()``` and s```of_dai_unload``` since current load/unload sequence (```sof_dai_load()```  -> ```sof_pcm_new()``` and ```sof_dai_unload()``` -> ```sof_pcm_free()```) looks odd. What do you think?

File: sound/soc/sof/topology.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/topology.c]: @plbossart this one's going to be a bit tricky. let me look into it tomorrow.

**Comment by ranj063** [comment]
> @ranj063 I have a suggestion. We can remove `sof_pcm_free()` and do all the allocate/free on `sof_dai_load()` and s`of_dai_unload` since current load/unload sequence (`sof_dai_load()` -> `sof_pcm_new()` and `sof_dai_unload()` -> `sof_pcm_free()`) looks odd. What do you think?

@bardliao yes that makes sense. I'll take a stab at getting rid of pcm_new/pcm_free to make the allocate/free balanced between dai load and unload.

**Comment by plbossart** [comment]
> > @ranj063 I have a suggestion. We can remove `sof_pcm_free()` and do all the allocate/free on `sof_dai_load()` and s`of_dai_unload` since current load/unload sequence (`sof_dai_load()` -> `sof_pcm_new()` and `sof_dai_unload()` -> `sof_pcm_free()`) looks odd. What do you think?
> 
> @bardliao yes that makes sense. I'll take a stab at getting rid of pcm_new/pcm_free to make the allocate/free balanced between dai load and unload.

Just to check the idea, the dai_load is really a cpu_dai_load which indirectly can lead to the PCM DMA allocation, right? If yes we'd need solid comments to explain the design since people unfamiliar with DPCM will get lost in translation.

**Comment by ranj063** [comment]
> > > @ranj063 I have a suggestion. We can remove `sof_pcm_free()` and do all the allocate/free on `sof_dai_load()` and s`of_dai_unload` since current load/unload sequence (`sof_dai_load()` -> `sof_pcm_new()` and `sof_dai_unload()` -> `sof_pcm_free()`) looks odd. What do you think?
> > 
> > 
> > @bardliao yes that makes sense. I'll take a stab at getting rid of pcm_new/pcm_free to make the allocate/free balanced between dai load and unload.
> 
> Just to check the idea, the dai_load is really a cpu_dai_load which indirectly can lead to the PCM DMA allocation, right? If yes we'd need solid comments to explain the design since people unfamiliar with DPCM will get lost in translation.

Yes @plbossart, thats correct. Sure, I will make extra effort to explain things properly.

**Comment by ranj063** [comment]
@plbossart @mengdonglin I'm working on this today. ETA ww02.3 for v2

**Comment by ranj063** [comment]
@plbossart @bardliao @keyonjie i've updated this PR now. I am still working on testing it further. So please do not merge.

**Comment by ranj063** [comment]
@plbossart could you please help review the v2 of this pr?

File: sound/soc/sof/topology.c [hunk 3]
```diff
@@ -2748,15 +2784,6 @@ void snd_sof_free_topology(struct snd_sof_dev *sdev)
...
-	list_for_each_entry_safe(sroute, temp, &sdev->route_list, list) {
-		sof_route_remove(&sroute->route);
-
-		/* free sroute and its private data */
-		kfree(sroute->private);
-		kfree(sroute);
-	}
-
 	ret = snd_soc_tplg_component_remove(sdev->component,
 					    SND_SOC_TPLG_INDEX_ALL);
```
Comment (bardliao) [inline sound/soc/sof/topology.c:2790]: Do we need snd_soc_tplg_component_remove()?. It is already called by [soc_cleanup_card_resources](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/soc-core.c#L2200) and [__snd_soc_unregister_component](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/soc-core.c#L3281)

Comment (ranj063) [reply sound/soc/sof/topology.c:2790]: @bardliao you are right about that. But if I remove that there's no need for free_topology at all.

Comment (ranj063) [reply sound/soc/sof/topology.c:2790]: @bardliao as we discussed, I need to keep the call to snd_soc_tplg_component_remove() to guarantee that sdev->tplg_loaded is set to false only after this call.

File: sound/soc/soc-topology.c [hunk 4]
```diff
@@ -1143,6 +1159,12 @@ static int soc_tplg_dapm_graph_elems_load(struct soc_tplg *tplg,
 		hdr->index);
 
 	for (i = 0; i < count; i++) {
+		struct snd_soc_dapm_route *route;
+
+		route = kzalloc(sizeof(*route), GFP_KERNEL);
+		if (!route)
+			return -ENOMEM;
```
Comment (plbossart) [inline sound/soc/soc-topology.c:1182]: if you have a mid-air memory allocation issue, the error handling needs to also free all previously allocated routes in this loop.
I am so predictable I know.

File: sound/soc/sof/topology.c [hunk 5]
```diff
@@ -1777,6 +1777,24 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
...
+
+	sroute = dobj->private;
+	if (!sroute)
+		return 0;
+
+	/* free sroute and its private data */
+	kfree(sroute->private);
+	list_del(&sroute->list);
+	kfree(sroute);
+
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1793]: in the previous patch, we do this in remove_route:

list_del(&dobj->list);
kfree(route);

Is this correct?

Comment (ranj063) [reply sound/soc/sof/topology.c:1793]: @plbossart oh yes. this was an oversight on my part. Let me add this to route_unload. Thanks!

Comment (ranj063) [reply sound/soc/sof/topology.c:1793]: @plbossart sorry I misunderstood. remove_route() calls dapm_route_unload() which frees the SOF driver data. But these two in remove_route() handle the freeing of the dapm_route which is dynamically allocated now

File: sound/soc/soc-topology.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/soc-topology.c:1182]: @plbossart Thanks for pointing this out. I need to add a sticky note on my monitor to remind myself every time I do this :)

**Comment by plbossart** [comment]
>   	for (i = 0; i < count; i++) {
> +		struct snd_soc_dapm_route *route;
> +
> +		route = kzalloc(sizeof(*route), GFP_KERNEL);
> +		if (!route)
> +			return -ENOMEM;
>
> @plbossart <https://github.com/plbossart> Thanks for pointing this 
> out. I need to add a sticky note on my monitor to remind myself every 
> time I do this :)
>
maybe allocate all the routes at once outside the loop so that the error 
flow remains simple?

**Comment by ranj063** [comment]
> for (i = 0; i < count; i++) { +	struct snd_soc_dapm_route *route; + +	route = kzalloc(sizeof(*route), GFP_KERNEL); +	if (!route) +	return -ENOMEM; @plbossart <https://github.com/plbossart> Thanks for pointing this out. I need to add a sticky note on my monitor to remind myself every time I do this :)
> maybe allocate all the routes at once outside the loop so that the error flow remains simple?

@plbossart fixed now.

File: sound/soc/sof/pcm.c [hunk 6]
```diff
@@ -546,58 +537,19 @@ static int sof_pcm_new(struct snd_soc_pcm_runtime *rtd)
...
-
-	if (spcm->pcm.capture)
-		snd_dma_free_pages(&spcm->stream[SNDRV_PCM_STREAM_CAPTURE].page_table);
-
-	/*
-	 * no need to explicitly release preallocated memory,
-	 * snd_pcm_lib_preallocate_free_for_all() is called by the core
-	 */
+	/* free pre-allocated audio buffers */
+	snd_pcm_lib_preallocate_free_for_all(pcm);
```
Comment (plbossart) [inline sound/soc/sof/pcm.c]: that's not needed, there was a comment on this that you also removed

/*	
	 * no need to explicitly release preallocated memory,	
	 * snd_pcm_lib_preallocate_free_for_all() is called by the core	
	 */

File: sound/soc/sof/pcm.c [hunk 7]
```diff
@@ -546,58 +537,19 @@ static int sof_pcm_new(struct snd_soc_pcm_runtime *rtd)
...
-
-	/* TODO: assign channel maps from topology */
-
 	return ret;
 
 free_playback_tables:
 	if (spcm->pcm.playback)
 		snd_dma_free_pages(&spcm->stream[SNDRV_PCM_STREAM_PLAYBACK].page_table);
 
-free_playback_dma_buffer:
```
Comment (plbossart) [inline sound/soc/sof/pcm.c:566]: why do we still have the free_playback_tables() if the allocation is done somewhere else?

File: sound/soc/soc-topology.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/soc-topology.c:1182]: this seems awfully complicated. why would we want to allocate memory twice?

File: sound/soc/soc-topology.c [hunk 8]
```diff
@@ -1157,20 +1192,33 @@ static int soc_tplg_dapm_graph_elems_load(struct soc_tplg *tplg,
...
 		/* add route, but keep going if some fail */
-		snd_soc_dapm_add_routes(dapm, &route, 1);
+		snd_soc_dapm_add_routes(dapm, routes[i], 1);
 	}
 
+	/*
+	 * free pointer to routes. The dapm routes will be freed
+	 * when the routes are removed in remove_route().
+	 */
+	kfree(routes);
```
Comment (plbossart) [inline sound/soc/soc-topology.c:1239]: I don't get this either? Allocated a routes array and play with an offset to get the routes you need, not sure what you are trying to do here?

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/topology.c:1793]: ok

File: sound/soc/soc-topology.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/soc-topology.c:1182]: @plbossart I am not allocating memory twice. "routes" is a pointer to an array of pointers. And routes[i] is a pointer to snd_soc_dapm_route element. 

The reason I dont allocate them all in one go is because we delete them individually in remove_route().

File: sound/soc/soc-topology.c (hunk 8 above)
Comment (ranj063) [reply sound/soc/soc-topology.c:1239]: @plbossart the same reason here routes is the pointer to the array of pointers. Each individual snd_soc_dapm_route element route[i] is freed in remove_route(). routes is just a pointer to the array of pointers which isnt needed anymore.

File: sound/soc/sof/pcm.c (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/pcm.c]: @plbossart yes, I remember making that change myself. Then in that case I'll remove pcm_free as well.

File: sound/soc/sof/pcm.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/pcm.c:566]: @plbossart this should go. will fix it

Comment (ranj063) [reply sound/soc/sof/pcm.c:566]: @plbossart fixed now

File: sound/soc/soc-topology.c [hunk 9]
```diff
@@ -1157,20 +1192,33 @@ static int soc_tplg_dapm_graph_elems_load(struct soc_tplg *tplg,
 			SNDRV_CTL_ELEM_ID_NAME_MAXLEN)
 			return -EINVAL;
```
Comment (plbossart) [inline sound/soc/soc-topology.c:1206]: You still have a memory leak on returning -EINVAL, routes is not freed and stuff that was allocated but not used is lost as well since the route is not added yet.

Comment (ranj063) [reply sound/soc/soc-topology.c:1206]: @plbossart @bardliao have fixed all opens in V4.

File: sound/soc/sof/topology.c [hunk 10]
```diff
@@ -1871,14 +1874,61 @@ static int sof_dai_load(struct snd_soc_component *scomp, int index,
...
+	caps = &spcm->pcm.caps[stream];
+
+	/* allocate playback page table buffer */
+	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV, sdev->parent,
+				  PAGE_SIZE, &spcm->stream[stream].page_table);
+	if (ret < 0)
+		dev_err(sdev->dev, "error: can't alloc page table for %s %d\n",
+			caps->name, ret);
+
+	return ret;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: so we if we playback+capture we still return here? that looks like a mistake

File: include/sound/soc-topology.h [hunk 11]
```diff
@@ -44,6 +44,7 @@ enum snd_soc_dobj_type {
 	SND_SOC_DOBJ_DAI_LINK,
 	SND_SOC_DOBJ_CODEC_LINK,
 	SND_SOC_DOBJ_WIDGET,
+	SND_SOC_DOBJ_ROUTE,
```
Comment (plbossart) [inline include/sound/soc-topology.h]: we should really reorder/cleanup this list to make it clear which of the dobj have a match in include/uapi/sound/asoc.h  and which ones are internal to the topology code.

File: sound/soc/soc-topology.c [hunk 12]
```diff
@@ -1142,36 +1160,85 @@ static int soc_tplg_dapm_graph_elems_load(struct soc_tplg *tplg,
...
+		snd_soc_dapm_add_routes(dapm, routes[i], 1);
 	}
 
-	return 0;
+	goto ret;
+
+err:
+	for (i = 0; i < count ; i++)
+		kfree(routes[i]);
+ret:
```
Comment (plbossart) [inline sound/soc/soc-topology.c]: goto statements are already borderline horrible, but if you use a label that is the name as a variable that's one step too far. Use something like "out", or maybe remove the goto err with break; and retest the value of ret to conditionally free the memory.

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (ranj063) [reply sound/soc/sof/topology.c]: @plbossart yes, fixed now.

File: include/sound/soc-topology.h (hunk 11 above)
Comment (ranj063) [reply include/sound/soc-topology.h]: @plbossart i've re-ordered this list and renamed ROUTE to GRAPH to align with the topology header type. I suppose there will be a conflict with Bard's PR but we can fix it depending on which order the PR's get merged

File: sound/soc/soc-topology.c (hunk 12 above)
Comment (ranj063) [reply sound/soc/soc-topology.c]: done

**Comment by ranj063** [comment]
@plbossart thats my understanding. they have been used interchangeably and also together in certain places like in soc-topology.h
/* DAPM graph route element loading and unloading */

But a graph elem in topology is equivalent to a snd_soc_dapm_route. I renamed it to align with the topology headers which use graph instead of route.

**Comment by plbossart** [comment]
let's merge then and test. It's not like we'll be breaking anything that's not broken already...
@bardliao please rebase your PR to align with the enum reordering. Thanks!
